### PR TITLE
asm2wasm atomics

### PR DESCRIPTION
**Pull request by kripken**
This does the core conversion of asm.js atomics code to wasm in asm2wasm.

Annoyance: atomic stores return a value in asm.js but not in wasm, requires a temp local.

Also fixes a wasm-builder bug in the alignment of atomic loads (should be the load size, not result size).

This is blocked on the test passing the optimizer (as we check that on all .asm.js tests), which I think doesn't have atomics support yet (it aborts in vacuum, which has code for each known node type).

### CODE REVIEW

**Comment by kripken** [comment]
cc @dschuff

**Comment by kripken** [comment]
I can take a look at the missing optimizer stuff for atomics.

**Comment by dschuff** [comment]
Cool; I added it in a couple of obvious places but definitely not everywhere, and it hasn't been tested on real code yet.

**Comment by kripken** [comment]
Ok, that is enough for the optimizer to run on that testcase. Good enough for this PR I think.

Output looks ok to me, but I don't have a deep understanding of atomics in either asm.js or wasm ;) so I could easily have missed something.

I haven't tried to actually run code, not sure if there is a wasm impl to test on. The binaryen interpreter doesn't have support either - not sure if we want to add that.

**Comment by dschuff** [comment]
@binji Are there any notable differences between `Atomics.*` and their wasm equivalents that we should be aware of?

File: test/threads.asm.js [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+  var Atomics_xor=global.Atomics.xor;
+  var tempFloat = Math_fround(0);
+  const f0 = Math_fround(0);
+
+  function test() {
+   var $temp = 0;
+   $temp = (Atomics_load(HEAP32, 1229)|0);
+   $temp = (Atomics_load(HEAPU16, 1229)|0);
+   $temp = (Atomics_store(HEAP32, $temp>>2, 0)|0);
+   $temp = (Atomics_exchange(HEAP32, $temp>>2, 1)|0);
```
Comment (binji) [inline test/threads.asm.js]: Does this change work with smaller sized Atomic RMW operations too?

**Comment by binji** [comment]
> I haven't tried to actually run code, not sure if there is a wasm impl to test on.

Not yet, unfortunately. I am working on a spec interpreter change for it, but I'm no ocaml expert, so it's taking time. I was thinking I might add it to wabt first since I know how that works. :-)

> Are there any notable differences between Atomics.* and their wasm equivalents that we should be aware of?

Only that there are 64-bit atomics. Otherwise they are the same (by design).

File: test/threads.asm.js (hunk 1 above)
Comment (kripken) [reply test/threads.asm.js]: It should, yeah, I added a commit with tests for a few more smaller size examples.

Comment (binji) [reply test/threads.asm.js]: Thanks, looks good.

**Pull request merged**